Add the MAS client usage in iKinGazeCtrl

Nicogene opened the pull request:
This PR add the possibility to make communicate the IMU and iKinGazeCtrl using the `multipleanalogsensorclient`.

This is a first draft of implementation. Name of the option, implementation, and name of the port can be discussed here or f2f.

cc @marcoaccame

pattacini commented on src/modules/iKinGazeCtrl/include/iCub/utils.h:
```diff
@@ -150,10 +154,14 @@ class ExchangeData
     bool            saccadesOn;
     bool            neckPosCtrlOn;
     bool            stabilizationOn;
+    bool            useMASClient;
```
I'd personally avoid handling this option.

My preference goes to testing this branch carefully - especially in terms of stabilization performance that are based on the gyro readouts - and then merge it. Thus, no switch, no command-line option to manage two different modalities.

Nicogene replied:
The only problem here is to update all the conf files of the robots in order to run the MAS server that wraps the imu. It is absolutely doable but it has to be taken in account

pattacini replied:
We don't need to deal with other robots since we'll be testing the branch with ours and eventually merge it. Let's remove the new parameter, then.

pattacini replied:
Sorry @Nicogene I didn't catch the point that not all the IMU's expose the new API.
I would then stop this development until we'll get all the low level layers aligned.

pattacini commented on src/modules/iKinGazeCtrl/src/utils.cpp:
```diff
@@ -360,6 +363,52 @@ Vector ExchangeData::get_imu()
+            double ts;
+            ret.second =  iGyro->getThreeAxisGyroscopeStatus(0) == MAS_OK;
+            ret.second &= iGyro->getThreeAxisGyroscopeMeasure(0, gyro, ts);
+            gyro = CTRL_DEG2RAD * gyro;
```
```suggestion
            gyro *= CTRL_DEG2RAD;
```

pattacini commented on src/modules/iKinGazeCtrl/src/utils.cpp:
```diff
@@ -360,6 +363,52 @@ Vector ExchangeData::get_imu()
+        else {
+            double ts;
+            ret.second =  iGyro->getThreeAxisGyroscopeStatus(0) == MAS_OK;
+            ret.second &= iGyro->getThreeAxisGyroscopeMeasure(0, gyro, ts);
```
```suggestion
            ret.second = iGyro->getThreeAxisGyroscopeMeasure(0, gyro, ts);
```

`ret.second` is undefined before this operation, as far as I see.

Nicogene replied:
It is valued with the status of the sensor just the line before, I don't know if it is the case to check the status of the sensor every time (@traversaro).

pattacini replied:
I was blind 👀

pattacini commented on src/modules/iKinGazeCtrl/src/utils.cpp:
```diff
@@ -360,6 +363,52 @@ Vector ExchangeData::get_imu()
+        else {
+            double ts;
+            ret.second =  iAccel->getThreeAxisLinearAccelerometerStatus(0) == MAS_OK;
+            ret.second &= iGyro->getThreeAxisGyroscopeMeasure(0, accel, ts);
```
```suggestion
            ret.second = iGyro->getThreeAxisGyroscopeMeasure(0, accel, ts);
```

pattacini commented on src/modules/iKinGazeCtrl/src/controller.cpp:
```diff
@@ -837,7 +837,8 @@ void Controller::run()
         {
-            Vector gyro=CTRL_DEG2RAD*commData->get_imu().subVector(6,8);
+            Vector gyro{commData->get_gyro().first};
+            reliableGyro = commData->get_gyro().second;
```
Two things:
- you're calling here `get_gyro()` twice, hence two readings, unless the compiler will do some optimizations; wouldn't it be better to use a temp variable to avoid that?
- `reliableGyro` shouldn't be put equal to `false` only because of the network, as I suppose it'll be done by `getThreeAxisGyroscopeMeasure`; instead, it is declared `false` when a specific condition is met (seek for it in the code). Then, how will these operations interact with the intended strategy? Could you elaborate more on this?

Nicogene replied:
About the variable you are probably right, it is slightly better than do two different function calls.

> reliableGyro shouldn't be put equal to false only because of the network.

Giving a second glance maybe this part of code should be changed, probably all the computations and checks should be skipped if the MAS client doesn't receive any data, do you agree?

pattacini replied:
- 👍  for 1.
- I agree with you that it'd be better to rethink 2, taking into account the original switch-off condition (that was unrelated to the network status).

pattacini commented on src/modules/iKinGazeCtrl/src/controller.cpp:
```diff
@@ -861,7 +862,8 @@ void Controller::run()
     {
-        Vector gyro=CTRL_DEG2RAD*commData->get_imu().subVector(6,8);
+        Vector gyro{commData->get_gyro().first};
+        reliableGyro = commData->get_gyro().second;
```
Same comments as above.

pattacini commented on src/modules/iKinGazeCtrl/src/main.cpp:
```diff
@@ -633,12 +638,12 @@ class GazeModule: public RFModule
         while (Time::now()-t0<ping_robot_tmo)
         {
-            if (pDrv!=NULL)
+            if (pDrv!=nullptr)
```
```suggestion
            delete pDrv;
```

pattacini commented on src/modules/iKinGazeCtrl/src/main.cpp:
```diff
@@ -1225,18 +1233,38 @@ class GazeModule: public RFModule
-                return false;
+            if (commData.useMASClient) {
+                Property mas_conf{{"device", Value("multipleanalogsensorsclient")},
+                                  {"remote", Value("/"+commData.robotName+"/head/imurfe")},
```
Could you please also take care of updating the documentation?

Nicogene replied:
Sure we have to settle the name of the ports, the first idea was `/robot_name/part/imu` or `imurfe` in case the robot mounts the robot facial expression board like the purple

pattacini commented on src/modules/iKinGazeCtrl/src/main.cpp:
```diff
@@ -1225,18 +1233,38 @@ class GazeModule: public RFModule
+                                  {"remote", Value("/"+commData.robotName+"/head/imurfe")},
+                                  {"local",  Value("/"+commData.localStemName+"/head/imurfe:i")}};
+
+                if (!(mas_client.open(mas_conf)) ||
```
I personally don't like this, because we compact checks that shall be sequential in one single verification relying on the properties of the compiler (and hiding this detail). Better off sticking to two subsequent checks instead.

pattacini commented:
Completely forgot the possible usage of [imuFilter][1], which is a module that can be appended to the inertial data to do the filtering needed by `iKinGazeCtrl`.

@Nicogene how does this PR deal with that?

[1]: http://www.icub.org/doc/icub-main/group__imuFilter.html

Nicogene commented:
> Completely forgot about the possible usage of imuFilter, which is a module that can be appended to the inertial data to do the filtering needed by iKinGazeCtrl.
@Nicogene how does this PR deal with that?

I honestly forgot about the `imuFilter`, I should chat with @traversaro what can be the best way to deal with it in the MAS client/server paradigm

traversaro commented:
> I honestly forgot about the imuFilter, I should chat with @traversaro what can be the best way to deal with it in the MAS client/server paradigm

Indeed, I also forgot about it. It is possible to build a MAS-based imuFilter, but I am not sure that is the best solution. In any case a f2f meeting with @Nicogene is required at least to discuss a first idea, that could be implemented once MAS wrapper has bee implemented on all robots.

Nicogene commented:
Here is the TODO list that came out after f2f meeting with @pattacini and @traversaro, for uniformly and transparently adopt the MAS client-server on all the iCub's imus.

- [ ] Rethink the handle of `reliableGyro`flag, taking account of network errors.
- [ ] Change the `xsens` device driver in order to expose the MAS interfaces.
- [ ] Refactor the `imuFilter` in order to make it a device driver that acts both as MAS wrapper and MAS client (man-in-the-middle).
- [ ] Remove the `useMasClient` option in `iKinGazeCtrl`.
- [ ] Change the xml files in `robots-configuration` in order to open the imus throught the MAS wrapper
- [ ] Update documentation

Nicogene commented:
# Progress Update 09/08/2019

I tested this patch on the `iCubGenova02` that mounts both `imuBoschBNO055` and the same imu on the rfe.

1. `iKinGazeCtrl` **without** imu work like before (expected, but to be checked)
2.  `iKinGazeCtrl` actually can't use the data coming from the imu on the rfe despite this changes, because it has a different orientation respect the old one. Some changes are required in the `iKin` library (see https://github.com/robotology/icub-main/issues/570#issuecomment-519943192).
3. I run the device `imuBoschBNO055` through the MASServer and then i run the `iKinGazeCtrl`. It worked, but it was a little more instable(more oscillation) respect the old version (reading raw data form the port). Moreover I sometimes get these errors.

```
[ERROR]MultipleAnalogSensorsClient: no data received in the last 0.010315 seconds, timeout enabled.
[ERROR]MultipleAnalogSensorsClient: Sensor of type ThreeAxisGyroscopes with index 0 has non-MAS_OK status.

```

Here the xml files i used for testing:

- `head-inertial.xml`

``` xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE devices PUBLIC "-//YARP//DTD yarprobotinterface 3.0//EN" "http://www.yarp.it/DTD/yarprobotinterfaceV3.0.dtd">


 <device xmlns:xi="http://www.w3.org/2001/XInclude" name="head-imuBosch" type="imuBosch_BNO055">
      <param name="i2c">  /dev/bosch-i2c-imu   </param>
</device>
```

- `head-imu-bosch_wrapper.xml`

``` xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE devices PUBLIC "-//YARP//DTD yarprobotinterface 3.0//EN" "http://www.yarp.it/DTD/yarprobotinterfaceV3.0.dtd">


    <device xmlns:xi="http://www.w3.org/2001/XInclude" name="head-imu-bosch_wrapper" type="multipleanalogsensorsserver">
        <param name="period">       10                  </param>
        <param name="name">     /icub/head/imu       </param>


        <action phase="startup" level="5" type="attach">
            <paramlist name="networks">
                <!-- The name of the element can be any string (we use SetOfIMUs to better express its nature).
                     Its value must match the device name in the corresponding body_part-jx_y-inertials.xml file
                     or in body_part-ebX-inertials.xml -->
                <elem name="SetOfIMUs"> head-imuBosch </elem>
            </paramlist>
        </action>

        <action phase="shutdown" level="5" type="detach" />
    </device>
```

cc @pattacini , @traversaro

Nicogene commented:
Today I tested successfully the `iKinGazeCtrl` with `xsens` exposing the MAS interface on `iCubGenova01`. Everything worked smoothly, and in this case no oscillation has been observed.

Nicogene commented:
Today I successfully tested, on `iCubGenova02` the `iKinGazeCtrl` with the `imuFilter` after being made a device driver.

I found and fixed two bugs meanwhile, one in yarp(https://github.com/robotology/yarp/commit/f56ce655a046453df9bcaac3186777fe82536510) and one in the icub device of the imurfe(https://github.com/robotology/icub-main/commit/f1fe695d8506d97c500553196f9b52611d537b9e).

i can state that the `iKinGazeCtrl` works worse while using the data coming `imurfe` for [this reason](https://github.com/robotology/icub-main/issues/595) I think.

Software side I think that everything is ready from my side. Ready for review.

Here the xml files used for running the new imufiter:

`head-imuFilter.xml`

``` xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE devices PUBLIC "-//YARP//DTD yarprobotinterface 3.0//EN" "http://www.yarp.it/DTD/yarprobotinterfaceV3.0.dtd">


    <device xmlns:xi="http://www.w3.org/2001/XInclude" name="head-imuFilter" type="imuFilter">
        <param name="period">       10                  </param>
        <param name="name">     /icub/head/imuFilter       </param>


        <action phase="startup" level="5" type="attach">
            <paramlist name="networks">
                <!-- The name of the element can be any string (we use SetOfIMUs to better express its nature).
                     Its value must match the device name in the corresponding body_part-jx_y-inertials.xml file
                     or in body_part-ebX-inertials.xml -->
                <elem name="SetOfIMUs"> head-inertial</elem>
            </paramlist>
        </action>

        <action phase="shutdown" level="5" type="detach" />
    </device>
```
`head-imuFilter_wrapper.xml`
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE devices PUBLIC "-//YARP//DTD yarprobotinterface 3.0//EN" "http://www.yarp.it/DTD/yarprobotinterfaceV3.0.dtd">


    <device xmlns:xi="http://www.w3.org/2001/XInclude" name="head-imuFilter_wrapper" type="multipleanalogsensorsserver">
        <param name="period">       10                  </param>
        <param name="name">     /icub/head/imuFilter       </param>


        <action phase="startup" level="6" type="attach">
            <paramlist name="networks">
                <!-- The name of the element can be any string (we use SetOfIMUs to better express its nature).
                     Its value must match the device name in the corresponding body_part-jx_y-inertials.xml file
                     or in body_part-ebX-inertials.xml -->
                <elem name="SetOfIMUs"> head-imuFilter </elem>
            </paramlist>
        </action>

        <action phase="shutdown" level="6" type="detach" />
    </device>
```

Nicogene commented:
Additionally the `imuFilter` can be run from command line:

```
yarpdev --device imuFilter --subdevice multipleanalogsensorsclient --proxy-remote <MASServerName> --proxy-local <MASClientName>
```

Tested also this case

pattacini commented:
🚀 Very nice job @Nicogene, as usual 😉 

You should rebase on `icub-main/devel` as I've just merged #594 so that you'll get the latest additions to `iKin`.

I've got a few questions regarding the test of `imuFilter` functionalities, but we can deal with them when I'll be back if it's ok to leave the PR open 'till the end of August.

Nicogene commented:
> I've got a few questions regarding the test of imuFilter functionalities, but we can deal with them when I'll be back if it's ok to leave the PR open 'till the end of August.

Yes sure, meanwhile I will deal with the changes required in `robots-configuration` and in the documentation of `iKinGazeCtrl` and `imuFilter`.

Nicogene commented:
I have rebased on the latest devel and changed one little thing then push-forced on the branch, if you see [here](https://github.com/Nicogene/icub-main/blob/c9547369db38e3eb209f7e8263394229f223d974/src/modules/iKinGazeCtrl/src/main.cpp#L1181) the code in the branch is updated but unfortunately the PR remains out of date :disappointed: 

Anyone knows any tip to workaround issues like this one?

Nicogene commented:
I updated the documentation of `imuFilter` and I think it is the only one that has to be changed.
The work on this PR should be finished.

pattacini commented:
As discussed today, before merging this, we wanna:
- polish the iCub startup script to get rid of `imuFilter`;
- expose the analog client wrapper parameter regulating the network timeout from within the gaze controller's config files; I'd rather keep the config files unchanged and put a `40 ms` default timeout in the main module

Nicogene commented:
dd84ab7 get rid of `imuFilter` from the xml files.
 
7ad26aa  add the imu::timeout.

From my side we can merge it
